Error: error:0308010C:digital envelope routines::unsupported when using Node 17

### Description of the bug / feature
A Vaadin 14 application does not start is you are using Node 17

```
2021-11-02 16:05:41.297 ERROR 8489 --- [onPool-worker-1] dev-webpack                              : Webpack failed with the exception:
(node:8503) [DEP0111] DeprecationWarning: Access to process.binding('http_parser') is deprecated.
(Use `node --trace-deprecation ...` to show where the warning was created)
ℹ ｢wds｣: Project is running at http://localhost:62996/webpack-dev-server/
ℹ ｢wds｣: webpack output is served from /VAADIN/
ℹ ｢wds｣: Content not from webpack is served from /Users/artur/test/lts-test, src/main/webapp
node:internal/crypto/hash:67
  this[kHandle] = new _Hash(algorithm, xofLen);
                  ^

Error: error:0308010C:digital envelope routines::unsupported
    at new Hash (node:internal/crypto/hash:67:19)
    at Object.createHash (node:crypto:130:10)
    at module.exports (/Users/artur/test/lts-test/node_modules/webpack/lib/util/createHash.js:135:53)
    at NormalModule._initBuildHash (/Users/artur/test/lts-test/node_modules/webpack/lib/NormalModule.js:417:16)
    at handleParseError (/Users/artur/test/lts-test/node_modules/webpack/lib/NormalModule.js:471:10)
    at /Users/artur/test/lts-test/node_modules/webpack/lib/NormalModule.js:503:5
```


### Minimal reproducible example
1. vaadin init lts-test
2. cd lts-test
3. mvn

### Expected behavior
The app starts

### Versions:
    - Vaadin / Flow version:14.7.3

Adding
```
export NODE_OPTIONS=--openssl-legacy-provider
```
fixes the webpack startup issue but it still does not work because loading localhost:8080 results in

```
com.vaadin.flow.server.BootstrapException: Unable to read webpack stats file.
	at com.vaadin.flow.server.BootstrapHandler$BootstrapPageBuilder.setupFrameworkLibraries(BootstrapHandler.java:956) ~[flow-server-2.7.2.jar:2.7.2]
	at com.vaadin.flow.server.BootstrapHandler$BootstrapPageBuilder.setupDocumentHead(BootstrapHandler.java:833) ~[flow-server-2.7.2.jar:2.7.2]
	at com.vaadin.flow.server.BootstrapHandler$BootstrapPageBuilder.getBootstrapPage(BootstrapHandler.java:600) ~[flow-server-2.7.2.jar:2.7.2]
	at com.vaadin.flow.server.BootstrapHandler.synchronizedHandleRequest(BootstrapHandler.java:514) ~[flow-server-2.7.2.jar:2.7.2]
	at com.vaadin.flow.server.SynchronizedRequestHandler.handleRequest(SynchronizedRequestHandler.java:40) ~[flow-server-2.7.2.jar:2.7.2]
	at com.vaadin.flow.server.VaadinService.handleRequest(VaadinService.java:1559) ~[flow-server-2.7.2.jar:2.7.2]
	at com.vaadin.flow.server.VaadinServlet.service(VaadinServlet.java:252) ~[flow-server-2.7.2.jar:2.7.2]
	at com.vaadin.flow.spring.SpringServlet.service(SpringServlet.java:111) ~[vaadin-spring-12.4.0.jar:na]
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:733) ~[tomcat-embed-core-9.0.45.jar:4.0.FR]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:227) ~[tomcat-embed-core-9.0.45.jar:9.0.45]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162) ~[tomcat-embed-core-9.0.45.jar:9.0.45]
	at org.apache.catalina.core.ApplicationDispatcher.invoke(ApplicationDispatcher.java:710) ~[tomcat-embed-core-9.0.45.jar:9.0.45]
...
Caused by: java.net.ConnectException: Connection refused
	at java.base/sun.nio.ch.Net.pollConnect(Native Method) ~[na:na]
	at java.base/sun.nio.ch.Net.pollConnectNow(Net.java:672) ~[na:na]
	at java.base/sun.nio.ch.NioSocketImpl.timedFinishConnect(NioSocketImpl.java:542) ~[na:na]
	at java.base/sun.nio.ch.NioSocketImpl.connect(NioSocketImpl.java:597) ~[na:na]
	at java.base/java.net.Socket.connect(Socket.java:633) ~[na:na]
	at java.base/sun.net.NetworkClient.doConnect(NetworkClient.java:178) ~[na:na]
	at java.base/sun.net.www.http.HttpClient.openServer(HttpClient.java:498) ~[na:na]
	at java.base/sun.net.www.http.HttpClient.openServer(HttpClient.java:603) ~[na:na]
	at java.base/sun.net.www.http.HttpClient.<init>(HttpClient.java:246) ~[na:na]
	at java.base/sun.net.www.http.HttpClient.New(HttpClient.java:351) ~[na:na]
	at java.base/sun.net.www.http.HttpClient.New(HttpClient.java:373) ~[na:na]
	at java.base/sun.net.www.protocol.http.HttpURLConnection.getNewHttpClient(HttpURLConnection.java:1309) ~[na:na]
	at java.base/sun.net.www.protocol.http.HttpURLConnection.plainConnect0(HttpURLConnection.java:1242) ~[na:na]
	at java.base/sun.net.www.protocol.http.HttpURLConnection.plainConnect(HttpURLConnection.java:1128) ~[na:na]
	at java.base/sun.net.www.protocol.http.HttpURLConnection.connect(HttpURLConnection.java:1057) ~[na:na]
	at java.base/sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1665) ~[na:na]
	at java.base/sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1589) ~[na:na]
	at java.base/java.net.HttpURLConnection.getResponseCode(HttpURLConnection.java:529) ~[na:na]
	at com.vaadin.flow.server.frontend.FrontendUtils.getStatsAssetsByChunkName(FrontendUtils.java:580) ~[flow-server-2.7.2.jar:2.7.2]
```

The webpack server is running though and is responding on the very port that Java tries to connec to

--- a/flow-server/src/main/java/com/vaadin/flow/server/frontend/FrontendTools.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/frontend/FrontendTools.java
@@ -378,11 +378,12 @@ public class FrontendTools {
             return;
         }
         try {
-            List<String> nodeVersionCommand = new ArrayList<>();
-            nodeVersionCommand.add(doGetNodeExecutable());
-            nodeVersionCommand.add("--version"); // NOSONAR
-            FrontendVersion foundNodeVersion = FrontendUtils.getVersion("node",
-                    nodeVersionCommand);
+            FrontendVersion foundNodeVersion = getNodeVersion();
+            if (foundNodeVersion.getMajorVersion() == 17
+                    && foundNodeVersion.getMinorVersion() == 0) {
+                throw new IllegalStateException(
+                        "Node version 17.0 is incompatible with webpack 4. Please upgrade to Node 17.1 or newer, or downgrade to Node 16.");
+            }
             FrontendUtils.validateToolVersion("node", foundNodeVersion,
                     SUPPORTED_NODE_VERSION, SHOULD_WORK_NODE_VERSION);
         } catch (UnknownVersionException e) {
@@ -402,6 +403,16 @@ public class FrontendTools {
             getLogger().warn("Error checking if npm is new enough", e);
         }
 
+    }
+
+    /**
+     * Gets the version of the node executable.
+     */
+    public FrontendVersion getNodeVersion() throws UnknownVersionException {
+        List<String> nodeVersionCommand = new ArrayList<>();
+        nodeVersionCommand.add(doGetNodeExecutable());
+        nodeVersionCommand.add("--version"); // NOSONAR
+        return FrontendUtils.getVersion("node", nodeVersionCommand);
     }
 
     /**

--- a/flow-server/src/main/java/com/vaadin/flow/server/frontend/FrontendVersion.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/frontend/FrontendVersion.java
@@ -284,6 +284,13 @@ public class FrontendVersion
     }
 
     @Override
+    public String toString() {
+        return "FrontendVersion [" + "majorVersion=" + majorVersion
+                + ", minorVersion=" + minorVersion + ", revision=" + revision
+                + ", buildIdentifier=" + buildIdentifier + "]";
+    }
+
+    @Override
     public int hashCode() {
         return (majorVersion + "." + minorVersion + "." + revision + "."
                 + buildIdentifier).hashCode();

--- a/flow-server/src/test/java/com/vaadin/flow/server/frontend/FrontendToolsTest.java
+++ b/flow-server/src/test/java/com/vaadin/flow/server/frontend/FrontendToolsTest.java
@@ -42,6 +42,7 @@ import org.junit.Ignore;
 import org.junit.Rule;
 import org.junit.Test;
 import org.junit.rules.TemporaryFolder;
+import org.mockito.Mockito;
 
 import com.vaadin.flow.server.frontend.installer.Platform;
 import com.vaadin.flow.server.frontend.installer.ProxyConfig;
@@ -78,7 +79,7 @@ public class FrontendToolsTest {
     public void setup() throws IOException {
         baseDir = tmpDir.newFolder().getAbsolutePath();
         vaadinHomeDir = tmpDir.newFolder().getAbsolutePath();
-        tools = new FrontendTools(baseDir, () -> vaadinHomeDir);
+        tools = Mockito.spy(new FrontendTools(baseDir, () -> vaadinHomeDir));
     }
 
     @Test
@@ -250,6 +251,15 @@ public class FrontendToolsTest {
         tools.validateNodeAndNpmVersion();
 
         Assert.assertTrue(file.exists());
+    }
+
+    @Test
+    public void validateNodeAndNpmVersion_brokenNode17() throws Exception {
+        Mockito.when(tools.getNodeVersion())
+                .thenReturn(new FrontendVersion(17, 0));
+        Assert.assertThrows(IllegalStateException.class, () -> {
+            tools.validateNodeAndNpmVersion();
+        });
     }
 
     @Test(expected = IllegalStateException.class)

--- a/vaadin-dev-server/src/main/java/com/vaadin/base/devserver/DevModeHandlerImpl.java
+++ b/vaadin-dev-server/src/main/java/com/vaadin/base/devserver/DevModeHandlerImpl.java
@@ -59,6 +59,8 @@ import com.vaadin.flow.server.VaadinSession;
 import com.vaadin.flow.server.communication.StreamRequestHandler;
 import com.vaadin.flow.server.frontend.FrontendTools;
 import com.vaadin.flow.server.frontend.FrontendUtils;
+import com.vaadin.flow.server.frontend.FrontendVersion;
+import com.vaadin.flow.server.frontend.FrontendUtils.UnknownVersionException;
 import com.vaadin.flow.server.startup.ApplicationConfiguration;
 
 import static com.vaadin.flow.server.Constants.VAADIN_MAPPING;
@@ -655,6 +657,18 @@ public final class DevModeHandlerImpl
                 () -> FrontendUtils.getVaadinHomeDirectory().getAbsolutePath(),
                 useHomeNodeExec);
         tools.validateNodeAndNpmVersion();
+        FrontendVersion nodeVersion;
+        try {
+            nodeVersion = tools.getNodeVersion();
+        } catch (UnknownVersionException e) {
+            getLogger().error("Unable to determine node version", e);
+            // Need to assume something..
+            nodeVersion = new FrontendVersion(16, 0, 0);
+        }
+        if (!nodeVersion.isOlderThan(new FrontendVersion(17, 0, 0))) {
+            processBuilder.environment().put("NODE_OPTIONS",
+                    "--openssl-legacy-provider");
+        }
 
         String nodeExec = null;
         if (useHomeNodeExec) {

